### PR TITLE
feat: refactor megatron init

### PR DESCRIPTION
**Pull request by ashors1**
# What does this PR do ?

**Add a one line overview of what this PR aims to accomplish.**

# Issues
List issues that this PR closes ([syntax](https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)):

# Usage
* **You can potentially add a usage example below**

```python
# Add a code snippet demonstrating how to use this
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](/NVIDIA-NeMo/RL/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA-NeMo/RL/blob/main/docs/testing.md) for how to run tests
- [ ] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA-NeMo/RL/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information
* ...

## Release Notes

* **New Features**
  * Enhanced Megatron model setup infrastructure with improved configuration validation, runtime management, and support for advanced parallelism options

* **Refactor**
  * Streamlined model initialization by consolidating setup logic into a modular, staged workflow improving maintainability and reducing code coupling

* **Tests**
  * Added comprehensive test suite validating configuration, setup validation, training initialization, and edge cases

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: nemo_rl/models/megatron/setup.py [hunk 1]
```diff
@@ -0,0 +1,928 @@
...
+            print(
+                f"[WARNING] Resetting async calls queue with {num_unfinalized} unfinalized calls"
+            )
+        try:
+            nemo_async_utils._async_calls_queue.close()
+        except:
+            pass  # Ignore errors during cleanup
+        # Reset the global async calls queue by creating a new instance
+        nemo_async_utils._async_calls_queue = AsyncCallsQueue()
+        print(f"[DEBUG] Reset NeMo async calls queue (old call_idx: {old_call_idx})")
```
Comment (joyang-nv) [inline nemo_rl/models/megatron/setup.py]: Is this debug info intended to be left? :)

Comment (ashors1) [reply nemo_rl/models/megatron/setup.py]: Thanks for the catch, removed it

File: nemo_rl/models/policy/workers/megatron_policy_worker.py [hunk 2]
```diff
@@ -444,480 +185,89 @@ def __init__(
...
         **kwargs: Any,
     ):
-        self.is_generation_colocated = None
-        if "generation" in config and config["generation"] is not None:
-            self.is_generation_colocated = config["generation"]["colocated"]["enabled"]
-
-        # Explicitly set NCCL_CUMEM_ENABLE to 1 to avoid the P2P initialization error for PyNCCLCommunicator.
-        # See https://github.com/NVIDIA-NeMo/RL/issues/564 for more details.
-        if not self.is_generation_colocated:
-            os.environ["NCCL_CUMEM_ENABLE"] = "1"
```
Comment (yuki-97) [inline nemo_rl/models/policy/workers/megatron_policy_worker.py:524]: feels better to keep the comments in `setup.py`, so that people can know why we need to do this.

File: nemo_rl/models/megatron/setup.py [hunk 3]
```diff
@@ -0,0 +1,893 @@
...
+        module = self.module
+        # Handle VLM models where language model is nested
+        if hasattr(module, "language_model"):
+            module = module.language_model
+        if hasattr(module, "decoder") and hasattr(module.decoder, "layers"):
+            for layer in module.decoder.layers:
+                mlp = getattr(layer, "mlp", None)
+                router = getattr(mlp, "router", None) if mlp is not None else None
+                if router is not None and hasattr(
+                    router, "_maintain_float32_expert_bias"
```
Comment (terrykong) [inline nemo_rl/models/megatron/setup.py:990]: i know this wasn't introduced in this PR, but a Q: it's possible for the router to not have this? should we error more loudly?

File: nemo_rl/models/megatron/setup.py [hunk 4]
```diff
@@ -0,0 +1,893 @@
...
+        config["megatron_cfg"]["optimizer"]["use_distributed_optimizer"]
+        and config["megatron_cfg"]["distributed_data_parallel_config"][
+            "overlap_param_gather"
+        ]
+    )
+
+    return megatron_tokenizer, megatron_bridge, should_disable_forward_pre_hook, dp_size
+
+
+class CustomFloat16Module(Float16Module):
```
Comment (terrykong) [inline nemo_rl/models/megatron/setup.py]: should re rename this since it's purpose is to indicate it's for freezing the moe router?

File: nemo_rl/models/megatron/setup.py (hunk 3 above)
Comment (ashors1) [reply nemo_rl/models/megatron/setup.py:990]: I'm not sure who added this code initially, but maybe @yaoyu-33 or @yfw would know the answer?

File: nemo_rl/models/megatron/setup.py (hunk 4 above)
Comment (ashors1) [reply nemo_rl/models/megatron/setup.py]: Makes sense. Renamed it to `MoEFloat16Module`.

File: nemo_rl/models/policy/workers/megatron_policy_worker.py (hunk 2 above)
Comment (ashors1) [reply nemo_rl/models/policy/workers/megatron_policy_worker.py:524]: Good catch. I added the missing comments back to `setup.py`

File: nemo_rl/models/megatron/setup.py (hunk 3 above)
Comment (yfw) [reply nemo_rl/models/megatron/setup.py:990]: I don't think we currently have an implementation of a Megatron Router that does not have `_maintain_float32_expert_bias` (https://github.com/terrykong/Megatron-LM/blob/b73ae5cdab9d409fcface2b2f3c375710abe6911/megatron/core/transformer/moe/router.py#L204). If this function gets removed / renamed for some reason, I think it will fail loudly during refit so it is probably fine to not fail loudly here. There may also be some scenario in the future where we implement a Megatron Router that does not need to have this function called in which case it also shouldn't fail loudly. CC @ZhiyuLi-Nvidia
